src/v/datalake/partition_key_path.cc

```diff
@@ -0,0 +1,324 @@
+    }
+};
+
+ss::sstring escape(std::string_view s) {
+    static constexpr std::
+      array<std::pair<absl::string_view, absl::string_view>, 25>
+        encode_lut = {
+          {{"&", "%26"},   {"$", "%24"},  {"@", "%40"}, {"=", "%3D"},
+           {";", "%3B"},   {"/", "%2F"},  {":", "%3A"}, {"+", "%2B"},
+           {" ", "%20"},   {",", "%2C"},  {"?", "%3F"}, {"\"", "%22"},
+           {"#", "%23"},   {"%", "%25"},  {"<", "%3C"}, {">", "%3E"},
+           {"[", "%5B"},   {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
+           {"`", "%60"},   {"{", "%7B"},  {"|", "%7C"}, {"}", "%7D"},
+           {"\177", "%7F"}}};
+
+    return absl::StrReplaceAll(s, encode_lut);
+}
+
+checked<ss::sstring, partition_key_error> format_field_value(
+  const iceberg::transform& transform,
+  const std::optional<iceberg::value>& value) {
+    if (!value) {
+        return "null";
+    }
+    if (!std::holds_alternative<iceberg::primitive_value>(*value)) {
+        return partition_key_error(
+          "non primitive iceberg partition values are not supported");
+    }
+    auto res = std::visit(
+      transform_value_formatting_visitor{
+        std::get<iceberg::primitive_value>(*value)},
+      transform);
+    if (res.has_error()) {
+        return res.error();
+    }
+    return escape(res.value());
+}
+} // namespace
+
+checked<remote_path, partition_key_error> partition_key_to_path(
+  const iceberg::partition_spec& spec, const iceberg::partition_key& key) {
+    std::filesystem::path ret;
+
+    if (!key.val) {
+        return partition_key_error{"Partition key value is missing"};
+    }
+
+    if (spec.fields.size() != key.val->fields.size()) {
+        return partition_key_error{ssx::sformat(
+          "Partition key/spec mismatch.The key has {} fields, but the spec has "
+          "{} fields",
+          key.val->fields.size(),
+          spec.fields.size())};
+    }
+    size_t total_length = 0;
+    for (size_t i = 0; i < spec.fields.size(); i++) {
+        auto& field = spec.fields[i];
+        const auto& value = key.val->fields[i];
+        auto formatting_res = format_field_value(field.transform, value);
+        if (formatting_res.has_error()) {
+            return formatting_res.error();
+        }
+        auto element = ssx::sformat(
+          "{}={}", escape(field.name), formatting_res.value());
+        total_length += element.size() + 1;
+        if (total_length > max_path_length) {
+            static constexpr auto rate_limit = std::chrono::seconds(5);
+            static thread_local ss::logger::rate_limit rate(rate_limit);
+            vloglr(
+              datalake_log,
+              ss::log_level::warn,
+              rate,
+              "Partition key path {}, exceeds the maximum length of {} bytes, "
+              "truncating",
+              ret,
+              max_path_length);
+            break;
+        }
+
+        ret /= std::filesystem::path(std::move(element));
```

Comment:
could we prepend path with a marker which indicates truncation? 🤔 could prevent some security/compatibility surprises

Reply:
I think this is not great idea, as it will change everything, let's maybe ask product team